glommio/src/error.rs

```diff
@@ -3,9 +3,60 @@
 //
 // This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2020 Datadog, Inc.
 //
-use std::fmt;
+use crate::{channels, executor, sync};
+use std::fmt::{self, Debug};
 use std::os::unix::io::RawFd;
 use std::path::PathBuf;
+use thiserror::Error;
+
+pub type Result<T, V> = std::result::Result<T, GlommioError<V>>;
+
+#[derive(Debug)]
+pub enum ResourceType {
+    Semaphore { requested: u64, available: u64 },
+}
+
+impl fmt::Display for ResourceType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ResourceType::Semaphore {
+                requested,
+                available,
+            } => f.write_str(&format!(
+                "Semaphore requested {} but only {} available",
+                requested, available
+            )),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+/// Composite error type to encompass all error types glommio produces.
+pub enum GlommioError<T: Debug + 'static> {
+    #[error("IO error occurred: {0}")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Lock closed error: {0}")]
+    LockClosedError(#[from] sync::LockClosedError),
+
+    #[error("Channel error: {0}")]
+    ChannelError(#[from] channels::ChannelError<T>),
+
+    #[error("Queue not found error: {0}")]
+    QueueNotFoundError(#[from] executor::QueueNotFoundError),
```

Comment:
Isn't it cleaner to get rid of the original error and only use the user-visible error ?
if I understand this crate correctly this should be possible by adding a parameter to QueueNotFoundError

Reply:
Yeah, I mention that somewhere in the description -- I agree that it would be cleaner.

Reply:
sorry, missed that.
That's great!